kotlinx-coroutines-core/common/src/JobSupport.kt

```diff
@@ -520,8 +547,12 @@ public open class JobSupport constructor(active: Boolean) : Job, ChildJob, Paren
         return node
     }
 
-    private fun addLastAtomic(expect: Any, list: NodeList, node: JobNode) =
-        list.addLastIf(node) { this.state === expect }
+    private fun addLast(expectedState: Any, list: NodeList, node: JobNode): Boolean {
+        list.addLast(node)
+        if (this.state === expectedState) return true
+        node.dispose()
```

Comment:
Do we need to call `dispose` here? If the state was updated, then there is no references to the list -> no need to remove node from the list

Reply:
I guess that's not correct. Consider transition to `Finishing` state where the list reference is captured or  transition from `Finishing` to `Completed` state that includes invocation of completion handlers from the list after the state cas (`completeStateFinalization`).

And skipping dispose may lead to duplicate nodes added to the list (when we retry to add one in `invokeOnCompletion`).

Reply:
Unfortunately, the `LinkedList` algorithm does not support removing and re-adding the same node. Consider the following scenario:
1. T1 calls `addLastIf` in `Incomplete` state and started executing `list.addLast(node)`, stalls (looses CPU) mid-progress inside addition.
2. T2 calls `addLastIf`, too, sees non-finished add by T1, starts "helping" it. but gets stalled too.
3. T3 transitions the state from `Incomplete` to `Finishing`.
4. T1 completes its `addLastIf`, sees that state has changed to `Finishing` and calls `node.dispose()`, which removes the node from the list. 
5. T1 now returns from `addLastIf` to `invokeOnCompletion` and retries. It discovers that the state is now `Finishing` and calls `addLastIf` again.
6. T1's calls to `addLastIf` calls `list.addLast(node)`. The first thing this call does is that it resets `prev` and `next` pointers in Node to point to itself. Assume it stalls right after that. 
7. T2 resumes and continues helping the original addition to the node. However, `prev` and `next` pointers are reset to point to the node itself, this helping algorithm can do all sorts of damage or hang.

I suggest the following change to fix it:

1. Inline both usages of `addLastIf` into their call sites. This will make reasoning about it easier. The first call to `addLastIf` from line 493 is trivial since the state is already `Finishing` there and it can only transition to `Complete`, so it does not suffer from "we need to remove it" problem at all. It is Ok if it stays added to the list. 

2. The only problematic call site of `addListIf` is on the line 526 when the state is `Incomplete`. When you detect that the state has changed, you can actually check _how_ its changed. If it becomes `Complete`, then it is trivial as above (you can safely retry) ,but if it became `Finishishing`, then you can actually use a boolean flag meaning "we've already added the node to the list", and check this flag before trying to add the node in the corresponding branches of `invokeOnCompletion` code to avoid adding the node the second time.


Instead of "removing the node to be readded later" I s